KNOX-3043 - NullpointerException in calling loadBuildProperties

senthh opened the pull request:
(It is very **important** that you created an Apache Knox JIRA for this change and that the PR title/commit message includes the Apache Knox JIRA ID!)

## What changes were proposed in this pull request?
Rootcause of this Issue:

Due to improper maven build, We were getting NullPointer Exception as Gatewayserver was not able to find 'build.properties' file (eventhough build.properties was part of gateway-server.jar file)

Error Messgae:
{{
SEVERE: Failed to start gateway: java.lang.NullPointerException
java.lang.NullPointerException
at java.util.Properties$LineReader.readLine(Properties.java:434)
at java.util.Properties.load0(Properties.java:353)
at java.util.Properties.load(Properties.java:341)
at org.apache.knox.gateway.GatewayServer.loadBuildProperties(GatewayServer.java:306)
at org.apache.knox.gateway.GatewayServer.main(GatewayServer.java:160)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at org.apache.knox.gateway.launcher.Invoker.invokeMainMethod(Invoker.java:68)
at org.apache.knox.gateway.launcher.Invoker.invoke(Invoker.java:39)
at org.apache.knox.gateway.launcher.Command.run(Command.java:99)
at org.apache.knox.gateway.launcher.Launcher.run(Launcher.java:75)
at org.apache.knox.gateway.launcher.Launcher.main(Launcher.java:52)
}}

Changes : Added Exceptional handlings to handle NullPointerException thrown while calling loadBuildProperties method and also respective log messages


## How was this patch tested?

Tested  by adding exceptional handling and then tested, in local cluster, for better error messages.

Adding exeption handling helped to troubleshoot the issue sooner.

Please review [Knox Contributing Process](https://cwiki.apache.org/confluence/display/KNOX/Contribution+Process#ContributionProcess-GithubWorkflow) before opening a pull request.

senthh commented:
@pzampino Could you please review this PR?

lmccay reviewed (changes requested):
Hi Senthil - thanks for your PR. Are you able to reproduce the NPE in a unit test and show that it is fixed? Can you also add more information to the PR description, explain how you have tested it, etc? Thanks again!

senthh commented:
> Hi Senthil - thanks for your PR. Are you able to reproduce the NPE in a unit test and show that it is fixed? Can you also add more information to the PR description, explain how you have tested it, etc? Thanks again!



> Hi Senthil - thanks for your PR. Are you able to reproduce the NPE in a unit test and show that it is fixed? Can you also add more information to the PR description, explain how you have tested it, etc? Thanks again!

@lmccay Thank you. I have added more description about root cause of this issue in description. We were getting this issue due to improper build of gateway-server. and we feel having good exceptional handling and error messges would help to troubleshoot this issue quicker than we took.

And I am adding these changes just to help others to troubhleshoot the issue quicker. And this issue got fixed by building with proper maven version.

moresandeep commented on gateway-server/src/main/java/org/apache/knox/gateway/util/KnoxCLI.java:
```diff
@@ -2570,10 +2570,16 @@ public String getUsage() {
-    try(InputStream inputStream = KnoxCLI.class.getClassLoader().getResourceAsStream( "build.properties" )) {
-      properties.load(inputStream);
+    String BUILD_PROPERTY = "build.properties";
+    PrintStream out = System.out;
```
We already have `out` defined on line 142 so we don't have to do it here again.

senthh replied:
@moresandeep  Even I tried to reuse 'out' defined in line 142, but as loadBuildProperties is a static method, it is not allowing to use "public PrintStream out = System.out;" from line 142.

moresandeep reviewed (changes requested):
Thank you @senthh for the patch, patch looks great, I have a minor change if you can address it that would be great.

lmccay commented:
> @lmccay Thank you. I have added more description about root cause of this issue in description. We were getting this issue due to improper build of gateway-server. and we feel having good exceptional handling and error messages would have helped to troubleshoot this issue quicker than we took.
> 
> I am adding these changes just to help others to trouble-shoot the issue quicker. And this issue got fixed by building with proper maven version. And, as we dont have existing Unit test suite for this issue we could not find it by running Testsuites.
> 
> But we tested the changes by creating the gateway-server.jar again with wrong maven config and validated if the new changes are reflected or not.

Interesting. It seems to me that we should also address the NPE itself, right? A null check and proper error message would help debug there as well - don't you think?

senthh commented:
> > @lmccay Thank you. I have added more description about root cause of this issue in description. We were getting this issue due to improper build of gateway-server. and we feel having good exceptional handling and error messages would have helped to troubleshoot this issue quicker than we took.
> > I am adding these changes just to help others to trouble-shoot the issue quicker. And this issue got fixed by building with proper maven version. And, as we dont have existing Unit test suite for this issue we could not find it by running Testsuites.
> > But we tested the changes by creating the gateway-server.jar again with wrong maven config and validated if the new changes are reflected or not.
> 
> Interesting. It seems to me that we should also address the NPE itself, right? A null check and proper error message would help debug there as well - don't you think?

Yes I correct!!.

senthh commented:
@lmccay and @moresandeep Please let me know if any changes/suggestions can help in this PR.

lmccay commented:
> > > @lmccay Thank you. I have added more description about root cause of this issue in description. We were getting this issue due to improper build of gateway-server. and we feel having good exceptional handling and error messages would have helped to troubleshoot this issue quicker than we took.
> > > I am adding these changes just to help others to trouble-shoot the issue quicker. And this issue got fixed by building with proper maven version. And, as we dont have existing Unit test suite for this issue we could not find it by running Testsuites.
> > > But we tested the changes by creating the gateway-server.jar again with wrong maven config and validated if the new changes are reflected or not.
> > 
> > 
> > Interesting. It seems to me that we should also address the NPE itself, right? A null check and proper error message would help debug there as well - don't you think?
> 
> Yes correct!!.

Do 

> > > @lmccay Thank you. I have added more description about root cause of this issue in description. We were getting this issue due to improper build of gateway-server. and we feel having good exceptional handling and error messages would have helped to troubleshoot this issue quicker than we took.
> > > I am adding these changes just to help others to trouble-shoot the issue quicker. And this issue got fixed by building with proper maven version. And, as we dont have existing Unit test suite for this issue we could not find it by running Testsuites.
> > > But we tested the changes by creating the gateway-server.jar again with wrong maven config and validated if the new changes are reflected or not.
> > 
> > 
> > Interesting. It seems to me that we should also address the NPE itself, right? A null check and proper error message would help debug there as well - don't you think?
> 
> Yes correct!!.

@senthh - do you plan to fix the NPE itself?

lmccay commented:
> > > @lmccay Thank you. I have added more description about root cause of this issue in description. We were getting this issue due to improper build of gateway-server. and we feel having good exceptional handling and error messages would have helped to troubleshoot this issue quicker than we took.
> > > I am adding these changes just to help others to trouble-shoot the issue quicker. And this issue got fixed by building with proper maven version. And, as we dont have existing Unit test suite for this issue we could not find it by running Testsuites.
> > > But we tested the changes by creating the gateway-server.jar again with wrong maven config and validated if the new changes are reflected or not.
> > 
> > 
> > Interesting. It seems to me that we should also address the NPE itself, right? A null check and proper error message would help debug there as well - don't you think?
> 
> Yes correct!!.

Do you plan to address the NPE itself?

senthh commented:
> > > > @lmccay Thank you. I have added more description about root cause of this issue in description. We were getting this issue due to improper build of gateway-server. and we feel having good exceptional handling and error messages would have helped to troubleshoot this issue quicker than we took.
> > > > I am adding these changes just to help others to trouble-shoot the issue quicker. And this issue got fixed by building with proper maven version. And, as we dont have existing Unit test suite for this issue we could not find it by running Testsuites.
> > > > But we tested the changes by creating the gateway-server.jar again with wrong maven config and validated if the new changes are reflected or not.
> > > 
> > > 
> > > Interesting. It seems to me that we should also address the NPE itself, right? A null check and proper error message would help debug there as well - don't you think?
> > 
> > 
> > Yes correct!!.
> 
> Do you plan to address the NPE itself? UPDATE: nevermind, I see that you do address the NPE in GatewayServer:306

Yes I have handled NPE and also added log messages too

lmccay commented:
@senthh - thank you for your contribution and interest in Apache Knox.
I have merged your PR and it will be available in the next release!

senthh commented:
> @senthh - thank you for your contribution and interest in Apache Knox. I have merged your PR and it will be available in the next release!

Thanks @lmccay
